docker/daemon/daemon_dest.go

```diff
@@ -157,15 +157,26 @@ func (d *daemonImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobI
 		defer os.Remove(streamCopy.Name())
 		defer streamCopy.Close()
 
-		size, err := io.Copy(streamCopy, stream)
+		// If we need to compute a digest.
+		digester := digest.Canonical.New()
+		tee := io.TeeReader(stream, digester.Hash())
+
+		size, err := io.Copy(streamCopy, tee)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
 		_, err = streamCopy.Seek(0, os.SEEK_SET)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+
+		if inputInfo.Digest == "" {
+			// In some (hacky) cases a caller might not know the size, but will
+			// know the digest (where "know" == "have a fake digest").
```

Comment:
Because DiffIDs != Digests, this happens for us (when copying we just have the names of the layers that we're dealing with, but we pretend like we don't have the Digest even though the two should be the same).

I'm going to keep this hunk change, but we can discuss the comment later.